evalml/pipelines/components/transformers/imputers/simple_imputer.py

```diff
@@ -69,42 +64,33 @@ def fit(self, X, y=None):
         if set([lt.type_string for lt in X.ww.logical_types.values()]) == {
             "boolean",
             "categorical",
-        } and not all(
-            [
-                is_categorical_actually_boolean(X, col)
-                for col in X.ww.select("Categorical")
-            ],
-        ):
+        }:
             raise ValueError(
                 "SimpleImputer cannot handle dataframes with both boolean and categorical features.  Use Imputer instead.",
             )
 
         nan_ratio = X.isna().sum() / X.shape[0]
         self._all_null_cols = nan_ratio[nan_ratio == 1].index.tolist()
 
-        X, _ = drop_natural_language_columns(X)
+        # Keep track of the different types of data in X
+        self._all_null_cols = nan_ratio[nan_ratio == 1].index.tolist()
+        self._natural_language_cols = X.ww.select("NaturalLanguage").columns.to_list()
 
-        # Convert any boolean columns to IntegerNullable, but keep track of the columns so they can be converted back
-        self._boolean_cols = list(
-            X.ww.select(
-                include=["Boolean", "BooleanNullable"],
-                return_schema=True,
-            ).columns,
-        )
-        # Make sure we're tracking Categorical columns that should be boolean as well
-        self._boolean_cols.extend(
-            [
-                col
-                for col in X.ww.select("Categorical")
-                if is_categorical_actually_boolean(X, col)
-            ],
-        )
-        X = set_boolean_columns_to_integer(X)
```

Comment:
Explanation as to why we can stop converting boolean columns to integers:
- This originally began as converting booleans to categorical at `fit` inside the simple imputer in https://github.com/alteryx/evalml/pull/3324. Not 100% clear why this was needed, but I imagine it had to do with a pandas incompatibility with booleans.
- Then, we started converting all boolean columns to categorical at `transform` as well in https://github.com/alteryx/evalml/pull/3892 (which I think was done because we couldn't reinfer the columns as booleans when they were imputed as booleans because sklearn turns them into 1s and 0s)
- Eventually, turning bools to categories became converting to IntegerNullable in https://github.com/alteryx/evalml/pull/3959 to handle a pandas bug with imputing categorical booleans.
- Given that the sklearn simple imputer's behavior is to turn boolean columns to 1s and 0s when imputed, we've now essentially recreated that behavior outside of it, so we should be able to just pass the original boolean and boolean nullable columns into the sklearn imputer and handle turning them back to Booleans at the end like we are now.

Reply:
I'm slightly worried that part of the original reason for converting boolean columns was because sklearn will not fit or transform its imputer on data that is only the `bool` dtype. Converting to category or int64 doesnt solves that problem because sklearn will happily take in data of only those types. 

At one point we had the logic that I now added back in to return early and not fit the imputer on sklearn's imputer object if only `bool` columns are present, and I havent been able to confirm it, but it's possible that part of the reason we removed that logic is that it's problematic if nans get introduced in the test data - we would have not fit the component, causing errors.

Because of that, I added logic to store the information if the component was fit on fully `bool` data, and if the data at transform isn't fully `bool`, call fit on the component of only the `cols_to_impute`.

Reply:
Looks like we have to handle this edge case if I understood correctly:
![Screenshot 2023-03-02 at 2 13 26 PM](https://user-images.githubusercontent.com/15271480/222528713-f594f501-c9b2-41a9-bc3a-c4f89803cdab.png)

"Given that the sklearn simple imputer's behavior is to turn boolean columns to 1s and 0s when imputed, we've now essentially recreated that behavior outside of it, so we should be able to just pass the original boolean and boolean nullable columns into the sklearn imputer and handle turning them back to Booleans at the end like we are now."

This is only true when the impute strategy is mean. Above we set it to `most_frequent` and we error out. This is the case we need since we only use `most_frequent` with bool columns in `Imputer`.

Could we convert the column to a string categorical or keep the current behavior of converting to a integer just for the `most_frequent` strategy or your suggestion above? I'd still prefer the conversion over refitting and transforming on the component object.

Reply:
Agreed with Jeremy here - calling fit during transform opens us up to future bugs. We only call fit once, and only ever call it on the training data. With a call to `_component_obj.fit` during transform, we now allow the component to fit on validation/testing data and overwrite earlier component training.
A quick test shows what would happen in the current setup:
![Screen Shot 2023-03-03 at 9 23 37 AM](https://user-images.githubusercontent.com/60074782/222744904-27202aa5-c5b6-453e-9c54-d83f2be6fee6.png)

We will be better served converting the booleans to integers rather than categoricals for `most_frequent` to match the behavior for `mean`, and agreed that conversion is preferable to this refitting.

Reply:
Thank you both for keeping me honest here! I definitely missed the nuance that sklearn allows `bool` only data for "mean" and "median" strategies. I think there are two main questions to answer:

1. How should we handle "mean" and "median" strategies for the SimpleImputer? We only allow `knn` or `most_frequent` as strategies for boolean columns in the `Imputer` component, but we don't restrict that in our `SimpleImputer` component despite the fact that we actually get a `TypeConversionError` if you attempt to use `median` or `mean` with a boolean nullable column that contains nans and the result contains not whole numbers (the cases becca's examples show - this is on main and my branch). For now, I think we should ignore this behavior since fixing it should be given proper consideration and it's not reachable from automl. I'll make an issue. 
3. How should we handle the case when only `bool` dtypes are passed to the SimpleImputer (independent of impute strategy which only really works for bool cols when its `most_frequent`)? We need to find a way to fit the `_component_obj` at that point, and I think we should keep the changes as minimal as possible, and only convert those columns to `BooleanNullable` (which sklearn is fine with!) when there are only `bool` columns in the data. There's no need to convert other `bool` columns, and I think converting to string or numeric values just opens us up to unexpected behavior now that sklearn has so much support for the `boolean` dtype. I'll double check with all of the different types that there isn't any unexpected behavior, or course.

Really, I just wish sklearn didn't have this weird limitation. Seems to completely miss the exact use case we need it for which is different training and test data.

Reply:
Created issue for the TypeConversionError: https://github.com/alteryx/evalml/issues/4050

Reply:
sounds good for 1. and agree with the `BooleanNullable` transform as long we set it back to `bool` after imputation!

Reply:
Good call out on making sure it goes back to Boolean after. I thiiink that's what's happening but I'll confirm.

Reply:
ah, we're only converting to `boolean` dtype in `fit`; we just then return early in transform if all the columns are `bool` , so the types will stay the same!